src/Commands/Server/Tools/AzureProxyTool.cs

```diff
@@ -0,0 +1,511 @@
+    {
+        var toolsJson = await GetToolListJsonAsync(request, tool);
+        if (string.IsNullOrEmpty(toolsJson))
+        {
+            return await RootLearnModeAsync(request, intent, cancellationToken);
+        }
+
+        var learnResponse = new CallToolResponse
+        {
+            Content =
+            [
+                new Content {
+                    Type = "text",
+                    Text = $"""
+                        Here are the available command and their parameters for '{tool}' tool.
+                        If you do not find a suitable tool, run again with the "learn" argument and empty "tool" to get a list of available tools and their parameters.
+                        Next, identify the command you want to execute and run again with the "tool", "command", and "parameters" arguments.
+
+                        {toolsJson}
+                        """
+                }
+            ]
+        };
+
+        var response = learnResponse;
+        if (SupportsSampling(request.Server) && !string.IsNullOrWhiteSpace(intent))
+        {
+            var (commandName, parameters) = await GetCommandAndParametersFromIntentAsync(request, intent, tool, toolsJson, cancellationToken);
+            if (commandName != null)
+            {
+                response = await CommandModeAsync(request, intent, tool, commandName, parameters, cancellationToken);
+            }
+        }
+        return response;
+    }
+
+    private async Task<CallToolResponse> CommandModeAsync(RequestContext<CallToolRequestParams> request, string intent, string tool, string command, Dictionary<string, object?> parameters, CancellationToken cancellationToken)
+    {
+        IMcpClient? client;
+
+        try
+        {
+            var clientOptions = CreateClientOptions(request.Server);
+            client = await _mcpClientProvider.GetProviderClientAsync(tool, clientOptions);
+            if (client == null)
+            {
+                _logger.LogError("Failed to get provider client for tool: {Tool}", tool);
+                return await RootLearnModeAsync(request, intent, cancellationToken);
+            }
+        }
+        catch (Exception ex)
+        {
+            _logger.LogError(ex, "Exception thrown while getting provider client for tool: {Tool}", tool);
+            return await RootLearnModeAsync(request, intent, cancellationToken);
+        }
+
+        try
+        {
+            var progressToken = request.Params?.Meta?.ProgressToken;
+            if (progressToken != null)
+            {
+                await request.Server.NotifyProgressAsync(progressToken.Value,
+                new ProgressNotificationValue
+                {
+                    Progress = 0f,
+                    Message = $"Calling {tool} {command}...",
+                }, cancellationToken);
+            }
```

Comment:
nit; Would it make sense to consolidate progress reporting into a single function, since it is currently done in multiple locations?